src/ipss_cell.hpp

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+#include <utility>
+#include <random>
+#include <limits>
+
+#include <cell_group.hpp>
+#include <recipe.hpp>
+#include <ipss_cell_description.hpp>
+#include <util/unique_any.hpp>
+
+
+namespace arb {
+
+/// Cell implementation of inhomogeneous Poisson spike generator.
+/// Spikes are generated at a configurable sample rate. With a rate that can be
+/// varied across time. Rate can be piece wise constant or linearly interpolated
+/// per sample time step.
+/// All times supplied to the cell will be 'rounded' to the first higher multiple of
+/// of the supplied sample_delta.
+/// When interpolating the rate will be the averaged rate between the
+/// t and t+1. The return spike time will be t.
+/// The first rate supplied in the time-rate vector should be at or before the
+/// start time of the cell. The rate of the last supplied time-rate will be kept
+/// until the stop_time of the cell
+
+class ipss_cell   {
+public:
+    /// Constructor
+    ipss_cell(ipss_cell_description desc, cell_gid_type gid) :
+        gid_(gid), sample_delta_(desc.sample_delta),
+        start_step_(time_to_step_idx(desc.start_time, sample_delta_)),
+        stop_step_(time_to_step_idx(desc.stop_time, sample_delta_)),
+        current_step_(0), interpolate_(desc.interpolate),
+        generator_(gid_)
+    {
+        // For internal storage convert the rates from spikes / second to
+        // spikes per sample_delta_  and the time to step_idx from t = 0.0
+        for (auto entry : desc.rates_per_time) {
+            auto spikes_per_sample_delta = (entry.second / 1000.0) * sample_delta_;
+            auto time_step = time_to_step_idx(entry.first, sample_delta_);
+            rates_per_step_.push_back({ time_step, spikes_per_sample_delta });
+        }
+
+        // Add and extra rate entry at max long, copy of the last original entry
+        // This allows current and next iterator.
+        rates_per_step_.push_back({ -1, rates_per_step_.cend()->second });
```

Comment:
`rates_per_step.back().second`

In fact, it looks to me like this should be an error: `cend()` should reference one past the end, so this is an undefined call past the end of the vector